Mathlib/LinearAlgebra/Matrix/PosDef.lean

```diff
@@ -38,59 +43,222 @@ open scoped Matrix
+  · refine ⟨le_antisymm ?_ (hA.2 v), le_antisymm ?_ (hB.2 v)⟩
+    · rw [add_comm, add_eq_zero_iff_eq_neg] at h_sum
+      simpa only [h_sum, neg_nonneg] using hB.2 v
+    · simpa only [add_eq_zero_iff_eq_neg.mp h_sum, neg_nonneg] using hA.2 v
+  have aux : star v ⬝ᵥ mulVec (A - B) v = 0
+  · rw [sub_mulVec, dotProduct_sub, h_van.1, h_van.2, sub_zero]
+  rw [hv', dotProduct_smul, IsROrC.real_smul_eq_coe_mul, ← mul_zero ↑t] at aux
+  exact hv <| Matrix.dotProduct_star_self_eq_zero.mp <| mul_left_cancel₀
+    (IsROrC.ofReal_ne_zero.mpr ht) aux
+
+lemma eq_sqrt_of_sq_eq {B : Matrix n n 𝕜} (hB : PosSemidef B) (hAB : A ^ 2 = B) : A = hB.sqrt :=
+  hA.eq_of_sq_eq_sq hB.posSemidef_sqrt (hB.sq_sqrt.symm ▸ hAB)
+
+end sqrt
+
+end PosSemidef
+
 @[simp]
 theorem posSemidef_submatrix_equiv {M : Matrix n n R} (e : m ≃ n) :
     (M.submatrix e e).PosSemidef ↔ M.PosSemidef :=
   ⟨fun h => by simpa using h.submatrix e.symm, fun h => h.submatrix _⟩
 #align matrix.pos_semidef_submatrix_equiv Matrix.posSemidef_submatrix_equiv
 
 /-- The conjugate transpose of a matrix mulitplied by the matrix is positive semidefinite -/
-theorem posSemidef_conjTranspose_mul_self (A : Matrix m n R) : Matrix.PosSemidef (Aᴴ * A) := by
+theorem posSemidef_conjTranspose_mul_self (A : Matrix m n R) : PosSemidef (Aᴴ * A) := by
   refine ⟨isHermitian_transpose_mul_self _, fun x => ?_⟩
   rw [← mulVec_mulVec, dotProduct_mulVec, vecMul_conjTranspose, star_star]
   exact Finset.sum_nonneg fun i _ => star_mul_self_nonneg _
 
 /-- A matrix multiplied by its conjugate transpose is positive semidefinite -/
-theorem posSemidef_self_mul_conjTranspose (A : Matrix m n R) : Matrix.PosSemidef (A * Aᴴ) :=
+theorem posSemidef_self_mul_conjTranspose (A : Matrix m n R) : PosSemidef (A * Aᴴ) :=
   by simpa only [conjTranspose_conjTranspose] using posSemidef_conjTranspose_mul_self Aᴴ
 
-/-- The eigenvalues of a positive semi-definite matrix are non-negative -/
-lemma PosSemidef.eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜}
-    (hA : Matrix.PosSemidef A) (i : n) : 0 ≤ hA.1.eigenvalues i :=
-  (hA.re_dotProduct_nonneg _).trans_eq (hA.1.eigenvalues_eq _).symm
-
 lemma eigenvalues_conjTranspose_mul_self_nonneg (A : Matrix m n 𝕜) [DecidableEq n] (i : n) :
     0 ≤ (isHermitian_transpose_mul_self A).eigenvalues i :=
-  (Matrix.posSemidef_conjTranspose_mul_self _).eigenvalues_nonneg _
+  (posSemidef_conjTranspose_mul_self _).eigenvalues_nonneg _
 
 lemma eigenvalues_self_mul_conjTranspose_nonneg (A : Matrix m n 𝕜) [DecidableEq m] (i : m) :
     0 ≤ (isHermitian_mul_conjTranspose_self A).eigenvalues i :=
-  (Matrix.posSemidef_self_mul_conjTranspose _).eigenvalues_nonneg _
+  (posSemidef_self_mul_conjTranspose _).eigenvalues_nonneg _
+
+/-- A matrix is positive semidefinite if and only if it has the form `Bᴴ * B` for some `B`. -/
+lemma posSemidef_iff_eq_transpose_mul_self [DecidableEq n] {A : Matrix n n 𝕜} :
+    PosSemidef A ↔ ∃ (B : Matrix n n 𝕜), A = Bᴴ * B := by
```

Comment:
This could be a bigger TFAE:
1. PosSemidef A
2. ∃ B, A = Bᴴ * B
3. ∃ H, H.IsHermitian, A = H * H
4. ∃ P, P.PosSemidef, A = P * P
5. A.IsHermitian ∧ ∀ x ∈ spectrum 𝕜 A, 0 ≤ x -- or another way to say the eigenvalues are nonnegative.

If you would like to leave this for a future PR that's okay.

Reply:
Personally I don't think TFAE's have any value unless:
* None of the spellings are canonical (`PosSemidef A` clearly is)
* The TFAE makes the proof shorter

Reply:
(but I would be in favor of an `Iff` lemma for each of those bullets except the first)

Reply:
Eric, I fundamentally disagree with this. I think that our current tooling for TFAE is the problem. I think we at least need a way to name the things in the list so that they can be used with `List.TFAE.out`.

Reply:
Since there is evidently not a consensus here as to whether this edit is beneficial, I am going to leave it for a future PR.

Reply:
My claim is that TFAE's are no good as an API surface because:
* They are unergonomic to use (as you have to refer to the statements by index). This could indeed be addressed by using names, but I don't see ``foo_tfae.out `foo `bar`` being more ergonomic than `foo_iff_bar`.
* They end up with weird asymmetries. You can't possibly put every `foo_iff_$BAR` in the same tfae, as some will have import constraints forcing them to be placed in other files.

TFAEs definitely have their place in setting up APIs, but I claim they should be an implementation detail.

Reply:
Incidentally, I had a stab at coding up the equivalence 1 ↔ 5 from Jireh's suggestion as a freestanding lemma, and it turns out to be much harder than it deserves to be. There seems to be no API linking `Algebra.Spectrum` of a matrix with its eigenvalues at present, so formulations using `spectrum` would need a whole load of extra work elsewhere.

I tried using a formulation based on "IsHermitian.eigenvalues", but I couldn't find a good formulation, because we don't have a definition of "eigenvalues" for a matrix that isn't known to be Hermitian, and I don't know how to write "X and [something which is only well-defined given X]" as a single expression. The best I can do is
```
lemma posSemidef_iff_eigenvalues_nonneg [DecidableEq n] {A : Matrix n n 𝕜} (hA : IsHermitian A):
    PosSemidef A ↔ ∀ i : n, 0 ≤ hA.eigenvalues I := by ...
```
I'm happy to add that to the PR if it's considered useful.

Reply:
PS: I guess `PosSemidef A ↔ ∃ (h : IsHermitian A), ∀ i : n, 0 ≤ h.eigenvalues I` will do the job, although it seems a bit clumsy.

Reply:
David, you can just avoid the TFAE for now if you would like. Thanks for being willing to try it though. Although the connection between spectrum and eigenvalues *does* exist, it's [Module.End.hasEigenvalue_iff_mem_spectrum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Basic.html#Module.End.hasEigenvalue_iff_mem_spectrum)

But @eric-wieser TFAEs have tremendous value from a *readability* perspective. Even if you have other `↔` statements lying around. This alone is enough to merit writing them, even if the full TFAE only holds under special circumstances (i.e., with sufficiently strong type class assumptions).

Moreover, proving a big TFAE is generally *easier* because if you have `n` statements you often only need `n` distinct `tfae_have`s, whereas to prove `1 ↔ 2`, ..., `1 ↔ n`, you need to show `2 * (n - 1)` different implications. A TFAE with 3 equivalences is arguably of insufficient value to warrant it, but a TFAE with 5 or more I think is easily worthwhile.

Reply:
> Although the connection between spectrum and eigenvalues does exist, it's [Module.End.hasEigenvalue_iff_mem_spectrum](https://leanprover-community.github.io/mathlib4_docs/Mathlib/LinearAlgebra/Eigenspace/Basic.html#Module.End.hasEigenvalue_iff_mem_spectrum)

Yes, I saw that lemma too, but the problem is the next step: we don't have the implication that, for Hermitian `T`, `Module.End.hasEigenvalue T a` implies that `a` is in `IsHermitian.eigenvalues T` (I'm deliberately omitting conversions between matrix and linear-map types here). We have the implication the other way, [LinearMap.IsSymmetric.hasEigenvalue_eigenvalues](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/InnerProductSpace/Spectrum.html#LinearMap.IsSymmetric.hasEigenvalue_eigenvalues), but not the way just stated. It wouldn't be all that hard to add, but it's not there as of now.

Reply:
I started a [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/TFAE.20discussion/near/408668294) about the use of TFAE in Mathlib.